### PR TITLE
Activate jsonld validation: checks for nested graph

### PR DESCRIPTION
**Pull request by lrosenstrom**

### CODE REVIEW

File: rest/src/main/groovy/whelk/rest/api/Crud.groovy [hunk 1]
```diff
@@ -92,6 +96,7 @@ class Crud extends HttpServlet {
         vocabData = whelk.vocabData
         jsonld = whelk.jsonld
         search = new SearchUtils(whelk)
+        validator = JsonLdValidator.from(jsonld)
```
Comment (olovy) [inline rest/src/main/groovy/whelk/rest/api/Crud.groovy:99]: Need to stop processing if validation fails, i.e. return.
I think it's cleaner to throw and write error response at top level.

File: rest/src/main/groovy/whelk/rest/api/Crud.groovy [hunk 2]
```diff
@@ -499,8 +504,8 @@ class Crud extends HttpServlet {
         newDoc.normalizeUnicode()
         newDoc.trimStrings()
         newDoc.deepReplaceId(Document.BASE_URI.toString() + IdGenerator.generate())
-        // TODO https://jira.kb.se/browse/LXL-1263
         newDoc.setControlNumber(newDoc.getShortId())
+        validateJsonLd(newDoc, request, response)
```
Comment (olovy) [inline rest/src/main/groovy/whelk/rest/api/Crud.groovy:508]: Same as above

File: rest/src/main/groovy/whelk/rest/api/Crud.groovy [hunk 3]
```diff
@@ -548,6 +553,17 @@ class Crud extends HttpServlet {
...
     }
 
+    private void validateJsonLd(Document doc, HttpServletRequest request, HttpServletResponse response) {
+        List<JsonLdValidator.Error> errors = validator.validate(doc.data)
+        if (errors) {
+            def message = errors.collect { it.toStringWithPath() }
+            failedRequests.labels(request.getMethod(), request.getRequestURI(),
+                    HttpServletResponse.SC_BAD_REQUEST.toString()).inc()
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST,
+                    "Invalid JsonLd. Got errors:" + message)
```
Comment (olovy) [inline rest/src/main/groovy/whelk/rest/api/Crud.groovy:563]: Maybe return a structured response, i.e. JSON?
(inconsistent with rest of API though)

**Review by olovy — CHANGES_REQUESTED** [review]
Need to stop processing if validation fails.

File: rest/src/main/groovy/whelk/rest/api/Crud.groovy (hunk 1 above)
Comment (lrosenstrom) [reply rest/src/main/groovy/whelk/rest/api/Crud.groovy:99]: Oops!

File: rest/src/main/groovy/whelk/rest/api/Crud.groovy (hunk 3 above)
Comment (lrosenstrom) [reply rest/src/main/groovy/whelk/rest/api/Crud.groovy]: We discussed this as a possible future feature, but skip it for now.

**Pull request merged**